grpc-ballerina/Module.md

```diff
@@ -0,0 +1,328 @@
+
+### gRPC Communication Patterns
+The common communication pattern between client and server is simple request-response style communication. However, with gRPC, you can leverage different inter-process communication patterns other than the simple request-response pattern.
+This module supports four fundamental communication patterns used in gRPC-based applications: simple RPC(unary RPC), server streaming RPC, client streaming RPC, and bidirectional streaming RPC.
+
+#### Simple RPC (Unary RPC) 
+In this pattern, the client invokes a remote function of a server and sends a single request to the server. The server sends a single response in return to the client along with status details.
+
+```proto
+service HelloWorld {
+    rpc hello (google.protobuf.StringValue)
+          returns (google.protobuf.StringValue);
+}
+```
+##### Creating the Server
+The code snippet given below contains a service that sends a response to each request.
+
+```ballerina
+// The gRPC service is attached to the listener.
+service HelloWorld on new grpc:Listener(9090)  {
+   // The function accepts a string message.
+   remote function hello(string name) returns string|error {
+       // Send the response to the client.
+       return "Hi " + name + "! Greetings from gRPC service!");
+   }
+}
+```
+##### Creating the Client
+The code snippet given below calls the above service in a synchronized manner using an auto-generated Ballerina stub.
+
+```ballerina
+// Use ‘HelloWorldClient’ to execute the call in the synchronized mode.
+HelloWorldClient helloClient = check new("http://localhost:9090");
+
+// Call the service remote function using a client stub.
+string responseFromServer = check helloClient->hello("Ballerina");
+```
+
+#### Server streaming RPC
+In server-side streaming RPC, the sends back a sequence of responses after getting the client's request message. After sending all the server responses, the server marks the end of the stream by sending the server status details.
+Users can invoke in a non-blocking manner.
+
+```proto
+service HelloWorld {
+    rpc lotsOfReplies (google.protobuf.StringValue)
+          returns (stream google.protobuf.StringValue);
+}
+```
+##### Creating the Server
+The code snippet given below contains a service that sends a sequence of responses to each request.
+
+```ballerina
+// The gRPC service is attached to the listener.
+service HelloWorld on new grpc:Listener(9090) {
+   remote function lotsOfReplies(string name) returns stream<string, grpc:Error?> {
+       string[] greets = ["Hi " + name, "Welcome " + name];
+       // Send multiple responses to the client.
+       return greets.toStream();
+   }
+}
+```
+
+##### Creating the Client
+The code snippet given below calls the above service using the auto-generated Ballerina client stub and reads multiple server responses using a stream.
+Here the message stream is ended with a `()` value.
+
+```ballerina
+   // Client endpoint configurations.
```

Comment:
Shouldn't any ballerina code sample start with zero indentation?
Let's check other instances as well.

Reply:
this is because these client codes come inside the main function or other function.